File: FlinkJavaCode/src/main/java/yangkai/com/flinkjava/code/chapter2/StreamBatchModeWordCount.java
package yangkai.com.flinkjava.code.chapter2;

import java.util.Arrays;
import org.apache.flink.api.common.RuntimeExecutionMode;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

/**
 * @author bytedance
 */
public class StreamBatchModeWordCount {

  public static void main(String[] args) throws Exception {
    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

    env.setRuntimeMode(RuntimeExecutionMode.AUTOMATIC);  // 会根据数据源来自动调整模式

    DataStreamSource<String> linesDS = env.readTextFile("./data/words.txt");

    linesDS
        .flatMap((String line, Collector<String> collector) -> Arrays.stream(line.split(" "))
            .forEach(collector::collect))
        .returns(Types.STRING)
        .map(word -> Tuple2.of(word, 1L))
        .returns(Types.TUPLE(Types.STRING, Types.LONG))
        .keyBy(new KeySelector<Tuple2<String, Long>, String>() {
          @Override
          public String getKey(Tuple2<String, Long> tp) throws Exception {
            return tp.f0;
          }
        })
        .sum(1)
        .print();

    // 流式计算中需要最后执行execute方法
    env.execute();
  }

}

File: FlinkScalaCode/src/main/scala/flinkscala/code/chapter2/StreamBatchModeWordCount.scala
package flinkscala.code.chapter2

import org.apache.flink.api.common.RuntimeExecutionMode
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment

object StreamBatchModeWordCount {
  def main(args: Array[String]): Unit = {
    //1.创建环境
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setRuntimeMode(RuntimeExecutionMode.BATCH)
    //2.导入隐式转换，使用Scala API 时需要隐式转换来推断函数操作后的类型
    import org.apache.flink.streaming.api.scala._

    //3.读取文件
    val ds: DataStream[String] = env.readTextFile("./data/words.txt")

    //4.进行wordCount统计
    ds.flatMap(line=>{line.split(" ")})
      .map((_,1))
      .keyBy(_._1)
      .sum(1)
      .print()

    //5.最后使用execute 方法触发执行
    env.execute()

  }
}
